ambry-store/src/main/java/com/github/ambry/store/FileSpan.java

```diff
@@ -53,7 +54,14 @@ Offset getEndOffset() {
    * @return {@code true} if {@code offset} is in this {@link FileSpan} (start and end offsets are considered inclusive)
    */
   boolean inSpan(Offset offset) {
-    return offset.compareTo(startOffset) >= 0 && offset.compareTo(endOffset) <= 0;
+    return offset.compareTo(startOffset) >= 0 && offset.compareTo(endOffset) < 0;
```

Comment:
I feel like there shouldn't be issues here, but could you double check the queries that `PersistentIndex.findEntriesSince()` uses to make sure that the exclusive end offset is okay? I think those methods use `getCurrentEndOffset` which uses the end offset of the last index segment. Is IndexSegment.getEndOffset() inclusive or exclusive?

Reply:
I have to double check with the code. But I want to point out something here, that end offset of index segment or log segment are kinda different than the end offset in the filespan. 

Say if we have a 10 puts in a log segment and each put is 1000 bytes, then this start offset of this log segment is 18 (with header), and the end offset is 1000 * 10 + 18 = 10018, which is expected.  Of course, this log segment would have index segment(s). Let's say, now we want to search within [18, 10018], both inclusive, we can have a filespan with start offset = 18 and end offset = 10018. Wait, the end offset in filespan is not inclusive, would this be a problem? Actually no. The 10 Puts, their offsets are 18, 1018, 2018, 3018... 9018 respectively. Even if we don't include 10018, we effectively covers all 10 puts here.

As you can see, the end offset of a log segment is greater than the offset of *LAST RECORD* in the segment, so an exclusive end offset in filespan can still cover all the records. (No records would be a zero-byte record)

The is also quite convenient for boundaries. There are two examples I want to share here.
1. Say we have a DELETE index value, and we have to find the corresponding PUT for this key. Then we can call `index.findKey(key, new FileSpan(index.getStartOffset(), deleteValue.getOffset(), EnumSet.of(PersistentIndex.IndexEntryType.PUT))`. The filespan's start offset is the start offset of the index, and the end offset is the DELETE's offset. Since the end offset is not included, so the DELETE record will be excluded from the search.
2. Say in blobstore, we have `indexEndOffsetBeforeCheck` and we want to do a findKey up to that offset. If we take a log segment above as an example, the `indexEndOffsetBeforeCheck` would be 10018 now. If before we call findKey, there is another put being inserted, then this last put's offset would be 10018, and in findKey, it will be excluded.